src/Authorization/FitOnFhir.Authorization/Handlers/FitOnFhirAuthenticationHandler.cs

```diff
@@ -0,0 +1,119 @@
+            _authenticationConfiguration = EnsureArg.IsNotNull(authenticationConfiguration, nameof(authenticationConfiguration));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _issuers = new Dictionary<string, string>();
+        }
+
+        /// <inheritdoc/>
+        public async Task<bool> AuthenticateToken(HttpRequest request, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(request, nameof(request));
+
+            // Ensure the request has a valid bearer token and extract the token from the header.
+            if (!request.TryGetTokenStringFromAuthorizationHeader(JwtBearerDefaults.AuthenticationScheme, out string token))
+            {
+                _logger.LogError("The request Authorization header is invalid.");
+                return false;
+            }
+
+            // Read the token into the handler.
+            var handler = new JwtSecurityTokenHandler();
+            handler.MapInboundClaims = false;
+            JwtSecurityToken jwtSecurityToken;
+            try
+            {
+                jwtSecurityToken = handler.ReadJwtToken(token);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "The request JWT is malformed.");
+                return false;
+            }
+
+            // Find the correct authority from the list
+            if (!_issuers.TryGetValue(jwtSecurityToken.Issuer, out var authority))
+            {
+                _logger.LogInformation("Issuer {0} not found in the list of authorized identity providers.", jwtSecurityToken.Issuer);
+                return false;
+            }
+
+            // Fetch the signing keys from the authority.
+            var config = await _openIdConfigurationProvider.GetConfigurationAsync(authority, cancellationToken);
+
+            try
+            {
+                var tokenValidationParameters = new TokenValidationParameters
+                {
+                    RequireSignedTokens = true,
+                    ValidAudience = _authenticationConfiguration.Audience,
+                    ValidateAudience = true,
+                    ValidIssuer = config.Issuer,
+                    ValidateIssuer = true,
+                    ValidateIssuerSigningKey = true,
+                    ValidateLifetime = true,
+                    IssuerSigningKeys = config.SigningKeys,
+                };
+
+                // Validate the token.
+                var tokenValidationResult = await handler.ValidateTokenAsync(jwtSecurityToken.RawData, tokenValidationParameters);
+
+                return tokenValidationResult.IsValid;
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Failed to validate the request Bearer token.");
+                return false;
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task CreateIssuerMapping(CancellationToken cancellationToken)
```

Comment:
I actually think this should be a private method and not expose the inner workings of this class to callers, and we should probably make this method non-async by using [TaskCompletionSource](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource-1?view=net-6.0), so you can call it from the constructor.

The gotcha with making this method non-async and calling it from the constructor is that the requests may not complete before the first call is made to AuthenticateToken(). This is mitigated by using TaskCompletionSource, because the AuthenticateToken() method can await TaskCompletionSource. In other words the long running process that is started by the constructor can then be awaited in another method.